src/Ledger/Prelude.agda

```diff
@@ -127,3 +129,23 @@ module Properties where
 
 _ᶠᵐ : {A B : Set} → A ⇀ B → FinMap A B
 (R , uniq) ᶠᵐ = (R , uniq , finiteness _)
+
+open import Interface.Hashable
+
+dupIsLeftUnique : ∀ {A : Set} → {x : ℙ A} → left-unique (map (λ x₁ → x₁ , x₁) x)
+dupIsLeftUnique {A} {x} x₁ x₂ with Properties.∈-map⁻' x₁ | Properties.∈-map⁻' x₂
+... | fst , refl , snd | .fst , refl , snd₁ = refl
+
+mapˡ-inj-dup :  ∀ {A B : Set} → (S : ℙ A)
+                            → (f : A → B)
+                            → Injective _≡_ _≡_ f
+                            → left-unique (mapˡ f (map (λ x → x , x) S))
+mapˡ-inj-dup S f x x₁ x₂ = mapˡ-uniq (λ _ _ → x) dupIsLeftUnique x₁ x₂
+
+setToHashRel : ∀ {A B : Set} → {{Hashable A B}} -> ℙ A -> ℙ (B × A)
+setToHashRel x = mapˡ hash (map (λ x₁ → x₁ , x₁) x)
+
+setToHashMap : ∀ {A B : Set} → {{Hashable A B}} -> ℙ A → B ⇀ A
+setToHashMap x with mapˡ-inj-dup x hash hashInj
+... | ans  = _ᵐ (setToHashRel x) {{record { isLeftUnique = λ x₁ x₂ → ans x₁ x₂ }}}
```

Comment:
This whole construction is not really necessary for the purposes of this PR; looking up a script hash (in `Ledger.ScriptValidation`) can be done directly like so:
```agda
scriptHashInTx : ScriptHash → Tx → Bool
scriptHashInTx sh tx = ⌊ any? (λ s → hash s ≟ sh) (tx .Tx.wits .TxWitnesses.scripts) ⌋
```

However, should we keep the "dup" construction (and the concrete use case that uses Hashable to pair elements with their hash), e.g. if you already have a use for it down the road to avoid people re-implementing all over again, or even without a use case in mind just to have it lying there under `Axiom.Set.Map`?
@WhatisRT WDYT?

Reply:
I agree it's not necessary, but it may be nice for the PDF. I don't recall this particular function in the Alonzo spec, so maybe it's a helper, but there are a couple of places where we do lookup things by their hashes, and writing `lookup h (setToHashMap s)` is much clearer than using `any?`.

That being said, I prefer to keep the set theory part separate from the ledger stuff (I want it to be its own library at some point). Maybe we could have a separate branch in the module tree for ledger-specific set theory stuff, like `Ledger.Set.X`?

Reply:
Just to clarify, this is what would be added in the set theory:
```agda
map⦅×-dup⦆-uniq : ∀ {x : Set A} → left-unique (mapˢ ×-dup x)
map⦅×-dup⦆-uniq x y with ∈-map⁻' x | ∈-map⁻' y
... | fst , refl , _ | .fst , refl , _ = refl

mapˡ∘map⦅×-dup⦆-uniq : ∀ {S : Set A} {f : A → B}
  → Injective _≡_ _≡_ f
  → left-unique $ mapˡ f (mapˢ ×-dup S)
mapˡ∘map⦅×-dup⦆-uniq f-inj = mapˡ-uniq (λ _ _ → f-inj) map⦅×-dup⦆-uniq
```

and then we just have to instantiate this to the hashing construction under Ledger/:
```agda
setToHashRel : ∀ {A B : Set} → {{Hashable A B}} -> ℙ A -> ℙ (B × A)
setToHashRel x = mapˡ hash (map ×-dup x)

setToHashMap : ∀ {A B : Set} → {{Hashable A B}} -> ℙ A → B ⇀ A
setToHashMap x = setToHashRel x ᵐ
  where instance _ = record { isLeftUnique = mapˡ∘map⦅×-dup⦆-uniq hashInj }
```

Or is it better to not even include the `×-dup` construction in `Axiom.Set.Map`?